### PR TITLE
[ESI] Introduce ESI pure module op

### PR DESCRIPTION
**Pull request by teqdruid**
A module containing only ESI channels and modules with only ESI ports. All non-local connectivity is done through ESI services. If this module is the top level in the design, then the design's actual top level ports are defined by a BSP.

Useful on its own for simulation and BSPs which don't define a top-level.

### CODE REVIEW

File: lib/Dialect/ESI/ESIOps.cpp [hunk 1]
```diff
@@ -422,6 +424,33 @@ void ServiceImplementReqOp::gatherPairedReqs(
...
+LogicalResult ESIPureModuleOp::verify() {
+  ESIDialect *esiDialect = getContext()->getLoadedDialect<ESIDialect>();
+
+  Block &body = getBody().front();
+  for (Operation &op : body.getOperations()) {
+    if (hw::HWInstanceLike inst = dyn_cast<hw::HWInstanceLike>(op)) {
+      for (auto operand : op.getOperands())
+        if (!operand.getType().isa<ChannelType>())
+          return inst.emitOpError(
+              "instances in ESI pure modules can only contain channel ports");
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIOps.cpp:440]: nit: use `llvm::any_of`.

File: lib/Dialect/ESI/ESIOps.cpp [hunk 2]
```diff
@@ -422,6 +424,33 @@ void ServiceImplementReqOp::gatherPairedReqs(
...
+      for (auto resType : op.getResultTypes())
+        if (!resType.isa<ChannelType>())
+          return inst.emitOpError(
+              "instances in ESI pure modules can only contain channel ports");
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIOps.cpp:444]: As above

**Review by mortbopet — COMMENTED** [review]
Is the assumption that clock/reset signals are always implicit for instantiated ops within this context? (because i assume we don't want to feed that through an ESI channel).

**Comment by teqdruid** [comment]
> Is the assumption that clock/reset signals are always implicit for instantiated ops within this context? (because i assume we don't want to feed that through an ESI channel).

Yes-ish. The plan for the future is that ESI modules would be able to request a clock/reset domain through a mechanism which looks a little bit like an ESI service. For now, though, clock/reset would be implicit and singular.

**Pull request merged**